[Permissions Syncs] add new metrics tracking

thenamankumar opened the pull request:
<img width="470" alt="image" src="https://user-images.githubusercontent.com/22571395/199183132-be95b46b-e6b2-4c1a-b19e-fea1c9a68f5d.png">


## Test plan

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

kopancek commented on enterprise/cmd/repo-updater/internal/authz/metrics.go:
```diff
@@ -45,9 +45,32 @@ var (
+		Name: "src_repoupdater_perms_syncer_perms_found",
+		Help: "The number of perms found for user/repo after sync",
+	}, []string{"type", "id"})
+	metricsPermsConsecutiveSyncDelay = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "src_repoupdater_perms_syncer_perms_consecutive_sync_delay",
+		Help: "The duration in minutes between last and current complete premissions sync.",
+	}, []string{"type"})
```
I think we talked on our zoom call, that this metric is basically defined per user/repo. So adding the id here similar to above would be IMO beneficial for analysis on which users/repos have the oldest perms.

kopancek commented on enterprise/cmd/repo-updater/internal/authz/metrics.go:
```diff
@@ -45,9 +45,32 @@ var (
+	metricsPermsFirstSyncDelay = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "src_repoupdater_perms_syncer_perms_first_sync_delay",
+		Help: "The duration in minutes it took for first user/repo complete perms sync after creation",
+	}, []string{"type"})
```
Same as above

kopancek commented on enterprise/cmd/repo-updater/internal/authz/metrics.go:
```diff
@@ -45,9 +45,32 @@ var (
+		Name: "src_repoupdater_perms_syncer_perms_first_sync_delay",
+		Help: "The duration in minutes it took for first user/repo complete perms sync after creation",
+	}, []string{"type"})
+	metricsItemsSyncScheduled = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "src_repoupdater_perms_syncer_items_sync_scheduled",
+		Help: "The number of users/repos scheduled for sync",
+	}, []string{"type", "priority"})
```
I think we can define this metric per user, repo (or add additional metric), similar to the metrics above for `src_repoupdater_perms_syncer_perms_found`. That way we can track if a specific repo/user gets scheduled for a sync.

I would also add the same type of metric (per repo, per user) for the following events:
- sync failure
- sync success

That way, we would be able to track the whole lifecycle of specific user or repo based on just metric data, without looking into the logs.

kopancek reviewed (commented):
Looks good to me.

Suggested some improvements for additional metrics.

thenamankumar commented:
@kopancek I updated all the metrics to include user/repo id and added a metric to track failed syncs as well so that we can track specific user perms syncing history solely from Prometheus.

kopancek commented on monitoring/definitions/repo_updater.go:
```diff
@@ -223,6 +223,167 @@ func RepoUpdater() *monitoring.Dashboard {
+							Description:    "number of first repo permissions syncs [5m]",
+                            Query:          `sum(increase(src_repoupdater_perms_syncer_initial_syncs{type="repo"}[5m]))`,
+							Panel:          monitoring.Panel().LegendFormat("{{type}}").Unit(monitoring.Number),
+							Owner:          monitoring.ObservableOwnerRepoManagement,
```
I just noticed this. We should change everything to IAM:
```suggestion
							Owner:          monitoring.ObservableOwnerIAM,
```

thenamankumar replied:
updated

kopancek commented on enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:
```diff
@@ -683,6 +696,16 @@ func (s *PermsSyncer) syncUserPerms(ctx context.Context, userID int32, noPerms b
+	metricsSuccessPermsSyncs.WithLabelValues("user", string(p.UserID)).Inc()
+	metricsPermsFound.WithLabelValues("user", string(p.UserID)).Set(float64(len(p.IDs)))
+
+	if &oldPerms.SyncedAt != nil {
```
`SyncedAt` is never going to be `nil`, so we will always end up not going in the other branch where we store the first sync metrics. See this code: https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@b104de5586c4ca6e1819ecb3d6e54351f06be81a/-/blob/enterprise/internal/database/perms_store.go?L1288&subtree=true

Same is true for the repositories.
```suggestion
	if !oldPerms.SyncedAt.IsZero() {
```

thenamankumar replied:
updated

kopancek commented on enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:
```diff
@@ -886,6 +917,17 @@ func (s *PermsSyncer) syncRepoPerms(ctx context.Context, repoID api.RepoID, noPe
+	metricsSuccessPermsSyncs.WithLabelValues("repo", string(p.RepoID)).Inc()
+	metricsPermsFound.WithLabelValues("repo", string(p.RepoID)).Set(float64(regularCount))
+
+	if &oldPerms.SyncedAt != nil {
```
```suggestion
	if !oldPerms.SyncedAt.IsZero() {
```
